### PR TITLE
Array agg groups accumulator

### LINKED ISSUE #10145: Implement GroupsAccumulator for array_agg aggregation function
### Is your feature request related to a problem or challenge?

Based on the article [Aggregating Millions of Groups Fast in Apache Arrow DataFusion](https://arrow.apache.org/blog/2023/08/05/datafusion_fast_grouping/) it's possible to optimise aggregation functions using GroupsAccumulator.

### Describe the solution you'd like

Implement GroupsAccumulator for array_agg aggregation function

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by lkt**
## Which issue does this PR close?

Closes #10145.

## Rationale for this change

See the issue.

## What changes are included in this PR?

GroupsAccumulator for `array_agg` aggregation function for:
  - Primitive types
  - String type

Not included:
  - Accumulating arrays of any level of nesting.

## Are these changes tested?

Extended tests in `array_agg.rs`

## Are there any user-facing changes?

No, just a performance improvement.

### CODE REVIEW

File: datafusion/physical-expr/src/aggregate/array_agg.rs [hunk 1]
```diff
@@ -96,6 +109,139 @@ impl AggregateExpr for ArrayAgg {
     fn name(&self) -> &str {
         &self.name
     }
+
+    fn groups_accumulator_supported(&self) -> bool {
+        self.input_data_type.is_primitive() || self.input_data_type == DataType::Utf8
```
Comment (Jefffrey) [inline datafusion/physical-expr/src/aggregate/array_agg.rs:114]: Are we able to support LargeUtf8 and Binary as well? Given how similar they are to Utf8?

**Review by Jefffrey — COMMENTED** [review]
Do we know of a benchmark we can run to test these changes? :eyes:

**Comment by alamb** [comment]
Thank you for this PR @lkt  -- I will take a look at this one carefully shortly

File: datafusion/physical-expr/src/aggregate/array_agg.rs [hunk 2]
```diff
@@ -96,6 +109,139 @@ impl AggregateExpr for ArrayAgg {
     fn name(&self) -> &str {
         &self.name
     }
+
+    fn groups_accumulator_supported(&self) -> bool {
```
Comment (comphead) [inline datafusion/physical-expr/src/aggregate/array_agg.rs:113]: I dont see this function used anywhere... 🤔

File: datafusion/physical-expr/src/aggregate/array_agg.rs [hunk 3]
```diff
@@ -187,17 +333,261 @@ impl Accumulator for ArrayAggAccumulator {
...
+    T: ArrowPrimitiveType + Send + Sync,
+{
+    fn update_batch(
+        &mut self,
+        new_values: &[ArrayRef],
+        group_indices: &[usize],
+        opt_filter: Option<&BooleanArray>,
+        total_num_groups: usize,
+    ) -> Result<()> {
+        assert_eq!(new_values.len(), 1, "single argument to update_batch");
```
Comment (comphead) [inline datafusion/physical-expr/src/aggregate/array_agg.rs:394]: Please return an internal_err!("") with more user friendly message

File: datafusion/physical-expr/src/aggregate/array_agg.rs [hunk 4]
```diff
@@ -187,17 +333,261 @@ impl Accumulator for ArrayAggAccumulator {
...
+
+impl GroupsAccumulator for StringArrayAggGroupsAccumulator {
+    fn update_batch(
+        &mut self,
+        new_values: &[ArrayRef],
+        group_indices: &[usize],
+        opt_filter: Option<&BooleanArray>,
+        total_num_groups: usize,
+    ) -> Result<()> {
+        assert_eq!(new_values.len(), 1, "single argument to update_batch");
```
Comment (comphead) [inline datafusion/physical-expr/src/aggregate/array_agg.rs:505]: ditto

File: datafusion/physical-expr/src/aggregate/array_agg.rs [hunk 5]
```diff
@@ -187,17 +333,261 @@ impl Accumulator for ArrayAggAccumulator {
...
+    }
+
+    fn merge_batch(
+        &mut self,
+        values: &[ArrayRef],
+        group_indices: &[usize],
+        opt_filter: Option<&BooleanArray>,
+        total_num_groups: usize,
+    ) -> Result<()> {
+        assert_eq!(values.len(), 1, "single argument to merge_batch");
```
Comment (comphead) [inline datafusion/physical-expr/src/aggregate/array_agg.rs:532]: ditto

**Comment by comphead** [comment]
Thanks @lkt for your contribution, does this PR stand for `array_agg` performance improvement? If so it is good to see a bench mark results

File: datafusion/physical-expr/src/aggregate/array_agg.rs (hunk 2 above)
Comment (lkt) [reply datafusion/physical-expr/src/aggregate/array_agg.rs:113]: It's called here datafusion/physical-plan/src/aggregates/row_hash.rs:393

File: datafusion/physical-expr/src/aggregate/array_agg.rs (hunk 3 above)
Comment (lkt) [reply datafusion/physical-expr/src/aggregate/array_agg.rs:394]: Thanks for your suggestion. I would keep the `assert_eq!` here to keep it consistent with other implementations of `GroupsAccumulator`. Eg. see here among others:
https://github.com/apache/datafusion/blob/main/datafusion/physical-expr/src/aggregate/groups_accumulator/prim_op.rs?plain=1#L93

Also this assertion is meant for developer to ensure the correct number of elements in the array between aggregation stages, not the end user. Let me know what you think.

Comment (comphead) [reply datafusion/physical-expr/src/aggregate/array_agg.rs:394]: Filed https://github.com/apache/datafusion/issues/10232

**Comment by alamb** [comment]
Seems to be superceded by https://github.com/apache/datafusion/pull/11096

**Pull request closed without merging**